### PR TITLE
BetonQuest (+ other plugin) support for non-persistent walls

### PR DESCRIPTION
**Pull request by Huskehhh**
Provides support for external plugins to utilise FakeBlock and create their own, non-persistent walls that can be controlled via code using List of UUID as opposed to permission node.

Targeted at #25 

What's left to do?
- [x] Revise implementation approach (should use an api instead? what would that design look like?)
- [x] Cleanup

It is likely this is implemented in place of an API (for now) however, will be reconsidered later in future.

### CODE REVIEW

File: intermediate/src/main/java/pro/husk/fakeblock/objects/IntermediateMaterialWall.java [hunk 1]
```diff
@@ -34,6 +35,39 @@ public IntermediateMaterialWall(String name, Location location1, Location locati
...
+
+        wallObject.setLocation1(location1);
+        wallObject.setLocation2(location2);
+
+        FakeBlock.newChain()
+                .async(() -> {
+                    wallObject.setFakeBlockDataHashMap(fakeBlockDataHashMap);
+                    wallObject.blocksInBetween.addAll(fakeBlockDataHashMap.keySet());
+                })
+                .sync(wallObject::removeOriginalBlocks)
```
Comment (Wolf2323) [inline intermediate/src/main/java/pro/husk/fakeblock/objects/IntermediateMaterialWall.java:61]: I think the behavior that the original blocks are removed is not wanted by every plugin. So what do you think about make this optional in this method? Also in the two ther MaterialWalls.
In BQ i do not prefere remove the original block.

**Review by Wolf2323 — COMMENTED** [review]
I like this it looks nice.
I left a comment about the original block remove.

One other thing is, if this behavior implemented that original blocks are not removed, we also may need a method, that redisplay the original blocks.

File: intermediate/src/main/java/pro/husk/fakeblock/objects/IntermediateMaterialWall.java (hunk 1 above)
Comment (Huskehhh) [reply intermediate/src/main/java/pro/husk/fakeblock/objects/IntermediateMaterialWall.java]: Fixed in latest commit

**Comment by Huskehhh** [comment]
> I like this it looks nice.
> I left a comment about the original block remove.
> 
> One other thing is, if this behavior implemented that original blocks are not removed, we also may need a method, that redisplay the original blocks.

There is currently this method:
https://github.com/Huskehhh/FakeBlock/blob/master/core/src/main/java/pro/husk/fakeblock/objects/WallObject.java#L181

However the functionality of this is to restore the state of the world prior to what it was when creating the Fake wall
Implementation at: 
https://github.com/Huskehhh/FakeBlock/blob/master/common/src/main/java/pro/husk/fakeblock/objects/CommonMaterialWall.java#L158

Consider these use case:

- Create a temporary, non persistent WallObject via added static methods (passing false as the replaceOriginalBlocks argument)
- Players who aren't on the list of WallObject will see the original state (as the world hasn't changed in any way)
- Players who are on the list of WallObject will see the desired altered state of the world (through fake packets)
- Server stops, no world changes were made so therefore can ignore cleanup of the fake wall

And on the contrary, another use case:
- Create a temporary, non persistent WallObject via added static methods (passing true as the replaceOriginalBlocks argument)
- Players who aren't on the list of WallObject will see the selected area as nothing (filled with air between location1 -> location2)
- Players who are on the list of WallObject will see the desired altered state of the world (through fake packets)
- Server stops, world changes are made, therefore cleanup is required if you do not wish to lose the state of the world.
- Calling restoreOriginalBlocks is NECESSARY to avoid loss of world data.

There's 2 approaches to this (that I can see as of now):
1. I can force WallObjects with the true replaceOriginalBlocks flag to be saved until they are cleaned up by the plugin interfacing with FakeBlock
2. Temporarily store the original block data, and on shutdown, restore the state.
3. Don't let temporary walls replace original block data, only fake the new state of it

Case 2 here seems like the worst solution as it runs the risk of data loss if the server shuts down for unknown reasons, as well as added memory impact from _more_ cached data.

**Comment by Huskehhh** [comment]
Took the approach of not allowing external users to replace the original blocks (unless they manually call it on the object), however, that is entirely up to them how they manage that! 👍 

This PR is likely complete and is (probably) ready for your testing

**Comment by SaltyAimbOtter** [comment]
Just wanted to let you know that Wolf will be back at this in a few days 😄 
He is currently busy writing his thesis.

**Comment by Huskehhh** [comment]
> Just wanted to let you know that Wolf will be back at this in a few days 😄
> He is currently busy writing his thesis.

Absolutely no dramas at all 👍

**Pull request merged**